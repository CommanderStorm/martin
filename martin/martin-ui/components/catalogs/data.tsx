"use client";

import {
	Database,
	Eye,
	Globe,
	ImageIcon,
	Layers,
	Palette,
	Search,
	Type,
} from "lucide-react";
import { ErrorState, InlineErrorState } from "@/components/error/error-state";
import { CatalogSkeleton } from "@/components/loading/catalog-skeleton";
import { LoadingSpinner } from "@/components/loading/loading-spinner";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import {
	Card,
	CardContent,
	CardDescription,
	CardHeader,
	CardTitle,
} from "@/components/ui/card";
import { Input } from "@/components/ui/input";
import { Tooltip, TooltipContent, TooltipTrigger } from "../ui/tooltip";

interface DataSource {
	id: string;
	name: string;
	type: string;
	description: string;
	layers: number;
	lastUpdated: string;
	size: string;
}

interface DataCatalogProps {
	dataSources: DataSource[];
	searchQuery: string;
	onSearchChangeAction: (query: string) => void;
	isLoading?: boolean;
	isSearching?: boolean;
	error?: Error | null;
	searchError?: Error | null;
	onRetry?: () => void;
	onRetrySearch?: () => void;
	isRetrying?: boolean;
}

export function DataCatalog({
	dataSources,
	searchQuery,
	onSearchChangeAction,
	isLoading = false,
	isSearching = false,
	error = null,
	searchError = null,
	onRetry,
	onRetrySearch,
	isRetrying = false,
}: DataCatalogProps) {
	if (isLoading) {
		return (
			<CatalogSkeleton
				title="Data Sources Catalog"
				description="Manage and explore all available tile sources, sprites, and fonts"
			/>
		);
	}

	if (error) {
		return (
			<ErrorState
				title="Failed to Load Data Catalog"
				description="Unable to fetch data sources from the server"
				error={error}
				onRetry={onRetry}
				isRetrying={isRetrying}
				variant="server"
				showDetails={true}
			/>
		);
	}

	const filteredDataSources = dataSources.filter(
		(source) =>
			source.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.type.toLowerCase().includes(searchQuery.toLowerCase()),
	);

	const getIcon = (type: string) => {
		switch (type) {
			case "Vector Tiles":
				return <Layers className="w-5 h-5 text-primary" />;
			case "Raster Tiles":
				return <ImageIcon className="w-5 h-5 text-primary" />;
			case "Sprites":
				return <Globe className="w-5 h-5 text-primary" />;
			case "Fonts":
				return <Type className="w-5 h-5 text-primary" />;
			default:
				return <Database className="w-5 h-5 text-primary" />;
		}
	};

	return (
		<div className="space-y-6">
			<div className="flex items-center justify-between">
				<div>
					<h2 className="text-2xl font-bold text-foreground">
						Data Sources Catalog
					</h2>
					<p className="text-muted-foreground">
						Manage and explore all available tile sources, sprites, and fonts
					</p>
				</div>
				<div className="relative">
					<Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 w-4 h-4" />
					<Input
						placeholder="Search data sources..."
						value={searchQuery}
						onChange={(e) => onSearchChangeAction(e.target.value)}
						className="pl-10 w-64 bg-card"
					/>
					{isSearching && (
						<div className="absolute right-3 top-1/2 transform -translate-y-1/2">
							<LoadingSpinner size="sm" />
						</div>
					)}
				</div>
			</div>

			{searchError && (
				<InlineErrorState
					message="Search failed. Please try again."
					onRetry={onRetrySearch}
					variant="network"
				/>
			)}

			{isSearching ? (
				<div className="flex items-center justify-center py-12">
					<div className="text-center">
						<LoadingSpinner size="lg" className="mx-auto mb-4" />
						<p className="text-muted-foreground">Searching data sources...</p>
					</div>
				</div>
			) : (
				<div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
					{filteredDataSources.map((source) => (
						<Card key={source.id} className="hover:shadow-lg transition-shadow">
							<CardHeader>
								<div className="flex items-center justify-between">
									<div className="flex items-center space-x-2">
										{getIcon(source.type)}
										<CardTitle className="text-lg">{source.name}</CardTitle>
									</div>
									<Badge variant="secondary">{source.type}</Badge>
								</div>
								<CardDescription>{source.description}</CardDescription>
							</CardHeader>
							<CardContent>
								<div className="space-y-2 text-sm text-muted-foreground">
									<div className="flex justify-between">
										<span>Layers:</span>
										<span>{source.layers}</span>
									</div>
									<div className="flex justify-between">
										<span>Size:</span>
										<span>{source.size}</span>
									</div>
									<div className="flex justify-between">
										<span>Updated:</span>
										<span>{source.lastUpdated}</span>
									</div>
								</div>
								<div className="flex space-x-2 mt-4">
  								<Tooltip>
                    <TooltipTrigger className="flex flex-1" asChild>
     									<Button
      										size="sm"
      										variant="outline"
      										className="flex-1 bg-transparent"
      										disabled
     									>
      										<Eye className="w-4 h-4 mr-2" />
      										Inspect
     									</Button>
                    </TooltipTrigger>
  									<TooltipContent>
                      <p>Not currently implemented in the frontend</p>
                    </TooltipContent>
                  </Tooltip>
  								<Tooltip>
                    <TooltipTrigger className="flex flex-1" asChild>
   									<Button
    										size="sm"
    										className="flex-1 bg-primary hover:bg-purple-700"
    										disabled
    								>
    										<Palette className="w-4 h-4 mr-2" />
    										Style
   									</Button>
    								</TooltipTrigger>
  									<TooltipContent>
                      <p>Not currently implemented in the frontend</p>
                    </TooltipContent>
                  </Tooltip>
								</div>
							</CardContent>
						</Card>
					))}
				</div>
			)}

			{!isSearching &&
				filteredDataSources.length === 0 &&
				searchQuery &&
				!searchError && (
					<div className="text-center py-12">
						<p className="text-muted-foreground">
							No data sources found matching "{searchQuery}"
						</p>
					</div>
				)}
		</div>
	);
}
