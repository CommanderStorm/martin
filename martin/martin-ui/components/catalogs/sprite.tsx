"use client";

import { Download, Eye, ImageIcon, Search } from "lucide-react";
import { ErrorState } from "@/components/error/error-state";
import { CatalogSkeleton } from "@/components/loading/catalog-skeleton";
import { LoadingSpinner } from "@/components/loading/loading-spinner";
import { SpriteDownloadModal } from "@/components/modals/sprite-download";
import { SpritePreviewModal } from "@/components/modals/sprite-preview";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import {
	Card,
	CardContent,
	CardDescription,
	CardHeader,
	CardTitle,
} from "@/components/ui/card";
import { Input } from "@/components/ui/input";

interface Sprite {
	name: string;
	id: string;
}

export interface SpriteCollection {
	name: string;
	description: string;
	icons: number;
	sizeInBytes: number;
	usage: string;
	sprites: Sprite[];
}

interface SpriteCatalogProps {
	selectedSprite: SpriteCollection | null;
	onSpriteSelectAction: (sprite: SpriteCollection) => void;
	onSpriteCloseAction: () => void;
	downloadSprite: SpriteCollection | null;
	onDownloadCloseAction: () => void;
	isLoading?: boolean;
	isLoadingSprites?: boolean;
	error?: string | Error | null;
	onRetry?: () => void;
	isRetrying?: boolean;
}

const spriteCollections: SpriteCollection[] = [
	{
		name: "POI Icons",
		description: "Point of interest markers and symbols",
		icons: 156,
		sizeInBytes: 2.3 * 1024 * 1024,
		usage: "45,230 requests/day",
		sprites: [
			{ name: "restaurant", id: "restaurant-icon" },
			{ name: "hotel", id: "hotel-icon" },
			{ name: "gas-station", id: "gas-station-icon" },
			{ name: "hospital", id: "hospital-icon" },
			{ name: "bank", id: "bank-icon" },
			{ name: "atm", id: "atm-icon" },
			{ name: "pharmacy", id: "pharmacy-icon" },
			{ name: "school", id: "school-icon" },
			{ name: "library", id: "library-icon" },
			{ name: "post-office", id: "post-office-icon" },
			{ name: "police", id: "police-icon" },
			{ name: "fire-station", id: "fire-station-icon" },
		],
	},
	{
		name: "Transportation",
		description: "Transit and transportation related icons",
		icons: 89,
		sizeInBytes: 1.8 * 1024 * 1024,
		usage: "32,180 requests/day",
		sprites: [
			{ name: "bus-stop", id: "bus-stop-icon" },
			{ name: "train-station", id: "train-station-icon" },
			{ name: "airport", id: "airport-icon" },
			{ name: "parking", id: "parking-icon" },
			{ name: "subway", id: "subway-icon" },
			{ name: "taxi", id: "taxi-icon" },
			{ name: "bicycle", id: "bicycle-icon" },
			{ name: "car-rental", id: "car-rental-icon" },
		],
	},
	{
		name: "Amenities",
		description: "Public amenities and services",
		icons: 124,
		sizeInBytes: 2.1 * 1024 * 1024,
		usage: "28,450 requests/day",
		sprites: [
			{ name: "wifi", id: "wifi-icon" },
			{ name: "restroom", id: "restroom-icon" },
			{ name: "information", id: "information-icon" },
			{ name: "wheelchair", id: "wheelchair-icon" },
			{ name: "elevator", id: "elevator-icon" },
			{ name: "stairs", id: "stairs-icon" },
			{ name: "drinking-water", id: "drinking-water-icon" },
			{ name: "phone", id: "phone-icon" },
		],
	},
	{
		name: "Recreation",
		description: "Parks, sports, and recreational facilities",
		icons: 67,
		sizeInBytes: 1.4 * 1024 * 1024,
		usage: "18,920 requests/day",
		sprites: [
			{ name: "park", id: "park-icon" },
			{ name: "playground", id: "playground-icon" },
			{ name: "stadium", id: "stadium-icon" },
			{ name: "beach", id: "beach-icon" },
			{ name: "swimming", id: "swimming-icon" },
			{ name: "tennis", id: "tennis-icon" },
			{ name: "golf", id: "golf-icon" },
			{ name: "hiking", id: "hiking-icon" },
		],
	},
	{
		name: "Shopping",
		description: "Retail and commercial establishments",
		icons: 78,
		sizeInBytes: 1.6 * 1024 * 1024,
		usage: "22,340 requests/day",
		sprites: [
			{ name: "shopping-mall", id: "shopping-mall-icon" },
			{ name: "grocery-store", id: "grocery-store-icon" },
			{ name: "clothing-store", id: "clothing-store-icon" },
			{ name: "electronics", id: "electronics-icon" },
			{ name: "bookstore", id: "bookstore-icon" },
			{ name: "flower-shop", id: "flower-shop-icon" },
			{ name: "jewelry", id: "jewelry-icon" },
			{ name: "bakery", id: "bakery-icon" },
		],
	},
	{
		name: "Custom Markers",
		description: "Custom branded location markers",
		icons: 24,
		sizeInBytes: 890 * 1024,
		usage: "12,670 requests/day",
		sprites: [
			{ name: "brand-a-marker", id: "brand-a-marker-icon" },
			{ name: "brand-b-marker", id: "brand-b-marker-icon" },
			{ name: "special-event", id: "special-event-icon" },
			{ name: "promotion", id: "promotion-icon" },
			{ name: "new-location", id: "new-location-icon" },
			{ name: "featured", id: "featured-icon" },
		],
	},
];

export function SpriteCatalog({
	selectedSprite,
	onSpriteSelectAction,
	onSpriteCloseAction,
	downloadSprite,
	onDownloadCloseAction,
	isLoading = false,
	isLoadingSprites = false,
	error = null,
	onRetry,
	isRetrying = false,
}: SpriteCatalogProps) {
	if (isLoading) {
		return (
			<CatalogSkeleton
				title="Sprite Catalog"
				description="Manage and preview all available sprite sheets and icons"
			/>
		);
	}

	if (error) {
		return (
			<ErrorState
				title="Failed to Load Sprites"
				description="Unable to fetch sprite catalog from the server"
				error={error}
				onRetry={onRetry}
				isRetrying={isRetrying}
				variant="server"
				showDetails={true}
			/>
		);
	}

	const handleSpriteSelect = (sprite: SpriteCollection) => {
		onSpriteSelectAction(sprite);
	};

	return (
		<>
			<div className="space-y-6">
				<div className="flex items-center justify-between">
					<div>
						<h2 className="text-2xl font-bold text-foreground">Sprite Catalog</h2>
						<p className="text-muted-foreground">
							Manage and preview all available sprite sheets and icons
						</p>
					</div>
					<div className="relative">
						<Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 w-4 h-4" />
						<Input placeholder="Search sprites..." className="pl-10 w-64" />
					</div>
				</div>

				<div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
					{spriteCollections.map((sprite: SpriteCollection, index: number) => (
						<Card key={index} className="hover:shadow-lg transition-shadow">
							<CardHeader>
								<div className="flex items-center justify-between">
									<div className="flex items-center space-x-2">
										<ImageIcon className="w-5 h-5 text-primary" />
										<CardTitle className="text-lg">{sprite.name}</CardTitle>
									</div>
									<Badge variant="secondary">1x, 2x</Badge>
								</div>
								<CardDescription>{sprite.description}</CardDescription>
							</CardHeader>
							<CardContent>
								<div className="space-y-4">
									<div className="p-3 bg-gray-50 rounded-lg">
										<p className="text-sm font-medium mb-2">Icon Preview:</p>
										<div className="grid grid-cols-8 gap-2">
											{Array.from({ length: Math.min(16, sprite.icons) }).map(
												(_: unknown, i: number) => (
													<div
														key={i}
														className="w-6 h-6 bg-purple-200 rounded flex items-center justify-center"
													>
														<div className="w-4 h-4 bg-primary rounded-sm"></div>
													</div>
												),
											)}
											{sprite.icons > 16 && (
												<div className="w-6 h-6 bg-gray-200 rounded flex items-center justify-center text-xs">
													+{sprite.icons - 16}
												</div>
											)}
										</div>
									</div>
									<div className="space-y-2 text-sm text-muted-foreground">
										<div className="flex justify-between">
											<span>Icons:</span>
											<span>{sprite.icons}</span>
										</div>
										<div className="flex justify-between">
											<span>File Size:</span>
											<span>{sprite.sizeInBytes} bytes</span>
										</div>
										<div className="flex justify-between">
											<span>Usage:</span>
											<span>{sprite.usage}</span>
										</div>
									</div>
									<div className="flex space-x-2">
										<Button
											variant="outline"
											size="sm"
											className="flex-1 bg-transparent"
										>
											<Download className="w-4 h-4 mr-2" />
											Download
										</Button>
										<Button
											variant="default"
											size="sm"
											className="flex-1 bg-primary hover:bg-purple-700"
											onClick={() => handleSpriteSelect(sprite)}
											disabled={isLoadingSprites}
										>
											{isLoadingSprites ? (
												<LoadingSpinner size="sm" className="mr-2" />
											) : (
												<Eye className="w-4 h-4 mr-2" />
											)}
											Preview
										</Button>
									</div>
								</div>
							</CardContent>
						</Card>
					))}
				</div>
			</div>

			<SpritePreviewModal
				sprite={selectedSprite}
				onCloseAction={onSpriteCloseAction}
				onDownloadAction={onSpriteSelectAction}
				isLoading={isLoadingSprites}
			/>

			<SpriteDownloadModal sprite={downloadSprite} onCloseAction={onDownloadCloseAction} />
		</>
	);
}
