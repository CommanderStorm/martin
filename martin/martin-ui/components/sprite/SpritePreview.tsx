import React, { useEffect, useState } from "react";
import { fetchSpriteImage, fetchSpriteIndex, SpriteMeta, SpriteIndex } from "./SpriteCache";
import SpriteCanvas from "./SpriteCanvas";
import { cn } from "@/lib/utils";

type SpritePreviewProps = {
  /**
   * Base URL for the sprite (without .json/.png or @2x).
   * Example: "https://example.com/sprite"
   */
  spriteUrl: string;
  /**
   * List of sprite IDs to display, in order. If not provided, all from the index are shown.
   */
  spriteIds: string[];
  /**
   * If true, only display the first PREVIEW_LIMIT sprites and use smaller icon size (for catalog previews).
   */
  previewMode?: boolean;
  /**
   * Optional className for the container.
   */
  className?: string;

};

type SpriteState =
  | { status: "loading" }
  | { status: "error"; error: string }
  | { status: "ready"; sprites: [string, SpriteMeta][]; image: HTMLImageElement };

export const SpritePreview: React.FC<SpritePreviewProps> = ({
  spriteUrl,
  spriteIds,
  previewMode,
  className,
}) => {
  const PREVIEW_LIMIT = 18;
  const [state, setState] = useState<SpriteState>({ status: "loading" });

  useEffect(() => {
    let cancelled = false;

    async function load() {
      setState({ status: "loading" });

      try {
        // we always use @2x high-DPI assets since we display them a little larger than one would on a map
        const [index, image] = await Promise.all([
          fetchSpriteIndex(`${spriteUrl}@2x.json`),
          fetchSpriteImage(`${spriteUrl}@2x.png`),
        ]);
        if (cancelled) return;

        let sprites = Object.entries(index);

        setState({
          status: "ready",
          sprites,
          image,
        });
      } catch (err: any) {
        if (cancelled) return;
        setState({
          status: "error",
          error: err?.message || "Failed to load sprite",
        });
      }
    }

    load();
    return () => {
      cancelled = true;
    };
  }, [spriteUrl]);

  // --- Determine which sprites to display ---
  let ids = spriteIds;
  if (previewMode && spriteIds.length > PREVIEW_LIMIT) {
    ids = ids.slice(0, PREVIEW_LIMIT - 1);
  }

  // --- Main grid of sprites ---
  if (state.status === "error") {
    return (
      <div className={`text-red-600 text-center p-4 ${className ?? ""}`}>
        <span>Error: {state.error}</span>
      </div>
    );
  }

  if (state.status === "ready" && state.sprites.length === 0) {
    return (
      <div className={`text-center p-4 ${className ?? ""}`}>
        <span>No sprites found.</span>
      </div>
    );
  }

  // Build a metaMap for fast lookup if ready
  const metaMap: Record<string, SpriteMeta> =
    state.status === "ready"
      ? Object.fromEntries(state.sprites)
      : {};

  return (
    <div
      className={cn(`flex flex-wrap gap-3 justify-start items-start min-h-[120px]`, className)}
    >
      {ids.map((id) =>
        <SpriteCanvas
          key={id}
          meta={metaMap[id]}
          image={state.status === "ready" ? state.image : undefined}
          label={id}
          previewMode={previewMode}
        />
      )}

      {/* +N bubble */}
      {previewMode && spriteIds.length > PREVIEW_LIMIT && (
        <div
          data-spritecnt = {spriteIds.length}
          className={`
            ${previewMode ? "h-7" : "h-12"}
            flex items-center justify-center
            ${previewMode ? "text-sm" : "text-lg"}
            text-gray-600 font-semibold
            ${previewMode ? "m-1.5" : "m-4"}
          `}
        >
          +{spriteIds.length - PREVIEW_LIMIT + 1}
        </div>
      )}
    </div>
  );
};

export default SpritePreview;
