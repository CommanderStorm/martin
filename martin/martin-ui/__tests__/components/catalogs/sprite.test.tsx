import { fireEvent, render, screen } from "@testing-library/react";
import type React from "react";
import { SpriteCatalog } from "@/components/catalogs/sprite";
import type { SpriteCollection } from "@/lib/types";

// Mock the components that use state instead of mocking React's useState
jest.mock("@/components/catalogs/sprite", () => {
  const originalModule = jest.requireActual("@/components/catalogs/sprite");
  const MockSpriteCatalog = (props: any) => {
    // Simple implementation that doesn't use useState
    const { spriteCollections, searchQuery, isLoading, error, onSearchChangeAction } = props;

    if (isLoading) {
      return (
        <div data-testid="catalog-skeleton">
          <div data-testid="skeleton-title">Sprite Catalog</div>
          <div data-testid="skeleton-description">
            Preview all available sprite sheets and icons
          </div>
        </div>
      );
    }

    if (error) {
      return (
        <div data-testid="error-state">
          <div data-testid="error-title">Failed to Load Sprites</div>
          <div data-testid="error-description">Unable to fetch sprite catalog from the server</div>
        </div>
      );
    }

    // Filter collections based on search
    const filteredCollections = Object.entries(spriteCollections || {}).filter(([name]) =>
      name.toLowerCase().includes(searchQuery.toLowerCase()),
    );

    return (
      <div>
        <h2>Sprite Catalog</h2>
        <div className="relative">
          <div data-testid="search-icon">Search</div>
          <input
            onChange={(e) => onSearchChangeAction(e.target.value)}
            placeholder="Search sprites..."
            value={searchQuery}
          />
        </div>

        <div className="grid">
          {filteredCollections.map(([name, sprite]: [string, any]) => (
            <div data-testid="card-header" key={name}>
              <div data-testid="card-title">{name}</div>
              <div data-testid="card-description">{sprite.images.length} total icons</div>
              <div data-testid="card-content">
                <div>{sprite.sizeInBytes / 1000} KB</div>
                <button
                  onClick={() => props.onPreviewClick && props.onPreviewClick(name)}
                >
                  <div data-testid="eye-icon">Eye</div>
                  Preview
                </button>
                <button onClick={() => props.onDownloadClick && props.onDownloadClick(name)}>
                  <div data-testid="download-icon">Download</div>
                  Download
                </button>
              </div>
            </div>
          ))}
        </div>

        {filteredCollections.length === 0 && searchQuery && (
          <div>No sprite collections found matching "{searchQuery}"</div>
        )}
      </div>
    );
  };

  return {
    ...originalModule,
    SpriteCatalog: MockSpriteCatalog,
  };
});

// Mock function handlers
const setSelectedSpriteMock = jest.fn();
const setDownloadSpriteMock = jest.fn();

// Mock all dependencies
jest.mock("@/components/error/error-state", () => ({
  ErrorState: ({ title, description }: { title: string; description: string }) => (
    <div data-testid="error-state">
      <div data-testid="error-title">{title}</div>
      <div data-testid="error-description">{description}</div>
    </div>
  ),
}));

jest.mock("@/components/loading/catalog-skeleton", () => ({
  CatalogSkeleton: ({ title, description }: { title: string; description: string }) => (
    <div data-testid="catalog-skeleton">
      <div data-testid="skeleton-title">{title}</div>
      <div data-testid="skeleton-description">{description}</div>
    </div>
  ),
}));

// Mock dialog components
jest.mock("@/components/dialogs/sprite-preview", () => ({
  SpritePreviewDialog: ({
    name,
    sprite,
    onCloseAction,
    onDownloadAction,
  }: {
    name: string;
    sprite: any;
    onCloseAction: () => void;
    onDownloadAction: () => void;
  }) => (
    <div data-testid="sprite-preview-dialog">
      <div data-testid="sprite-preview-name">{name}</div>
      <button data-testid="sprite-preview-close" onClick={onCloseAction}>
        Close
      </button>
      <button data-testid="sprite-preview-download" onClick={onDownloadAction}>
        Download
      </button>
    </div>
  ),
}));

jest.mock("@/components/dialogs/sprite-download", () => ({
  SpriteDownloadDialog: ({
    name,
    sprite,
    onCloseAction,
  }: {
    name: string;
    sprite: any;
    onCloseAction: () => void;
  }) => (
    <div data-testid="sprite-download-dialog">
      <div data-testid="sprite-download-name">{name}</div>
      <button data-testid="sprite-download-close" onClick={onCloseAction}>
        Close
      </button>
    </div>
  ),
}));

// Mock UI components to avoid tooltip provider issues
jest.mock("@/components/ui/tooltip", () => ({
  Tooltip: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  TooltipContent: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="tooltip-content">{children}</div>
  ),
  TooltipTrigger: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));

jest.mock("@/components/ui/button", () => ({
  Button: ({ children, onClick, ...props }: any) => (
    <button onClick={onClick} {...props}>
      {children}
    </button>
  ),
}));

jest.mock("@/components/ui/badge", () => ({
  Badge: ({ children, ...props }: any) => (
    <span data-testid="badge" {...props}>
      {children}
    </span>
  ),
}));

jest.mock("@/components/ui/input", () => ({
  Input: (props: any) => <input {...props} />,
}));

jest.mock("@/components/ui/card", () => ({
  Card: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  CardContent: ({ children, ...props }: any) => (
    <div data-testid="card-content" {...props}>
      {children}
    </div>
  ),
  CardDescription: ({ children, ...props }: any) => (
    <div data-testid="card-description" {...props}>
      {children}
    </div>
  ),
  CardHeader: ({ children, ...props }: any) => (
    <div data-testid="card-header" {...props}>
      {children}
    </div>
  ),
  CardTitle: ({ children, ...props }: any) => (
    <div data-testid="card-title" {...props}>
      {children}
    </div>
  ),
}));

// Mock utils
jest.mock("@/lib/utils", () => ({
  formatFileSize: (size: number) => `${size / 1000} KB`,
}));

jest.mock("lucide-react", () => ({
  Download: () => <div data-testid="download-icon">Download</div>,
  Eye: () => <div data-testid="eye-icon">Eye</div>,
  ImageIcon: () => <div data-testid="image-icon">Image</div>,
  Search: () => <div data-testid="search-icon">Search</div>,
}));

describe("SpriteCatalog Component", () => {
  const mockSpriteCollections: { [name: string]: SpriteCollection } = {
    "map-icons": {
      images: ["pin", "marker", "building", "park", "poi"],
      lastModifiedAt: new Date("2023-01-10"),
      sizeInBytes: 25000,
    },
    transportation: {
      images: ["car", "bus", "train", "bicycle", "walk", "plane", "ferry"],
      lastModifiedAt: new Date("2023-03-20"),
      sizeInBytes: 30000,
    },
    "ui-elements": {
      images: ["arrow", "plus", "minus", "close", "menu", "search", "filter", "settings"],
      lastModifiedAt: new Date("2023-02-15"),
      sizeInBytes: 35000,
    },
  };

  const defaultProps = {
    error: null,
    isLoading: false,
    onDownloadClick: jest.fn(),
    onPreviewClick: jest.fn(),
    onSearchChangeAction: jest.fn(),
    searchQuery: "",
    spriteCollections: mockSpriteCollections,
  };

  it("matches snapshot for loading state", () => {
    const { asFragment } = render(<SpriteCatalog {...defaultProps} isLoading={true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for loaded state with mock data", () => {
    const { asFragment } = render(<SpriteCatalog {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  // Reset mocks between tests
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders loading skeleton when isLoading is true", () => {
    render(<SpriteCatalog {...defaultProps} isLoading={true} />);
    expect(screen.getByTestId("catalog-skeleton")).toBeInTheDocument();
    expect(screen.getByTestId("skeleton-title").textContent).toBe("Sprite Catalog");
    expect(screen.getByTestId("skeleton-description").textContent).toBe(
      "Preview all available sprite sheets and icons",
    );
  });

  it("renders error state when there is an error", () => {
    const error = new Error("Test error");
    render(<SpriteCatalog {...defaultProps} error={error} />);
    expect(screen.getByTestId("error-state")).toBeInTheDocument();
    expect(screen.getByTestId("error-title").textContent).toBe("Failed to Load Sprites");
  });

  it("renders sprite collections correctly", () => {
    render(<SpriteCatalog {...defaultProps} />);
    expect(screen.getByText("Sprite Catalog")).toBeInTheDocument();

    // Get all card headers
    const headers = screen.getAllByTestId("card-header");
    expect(headers.length).toBe(3);

    // Check that each sprite collection name is displayed
    expect(screen.getByText("map-icons")).toBeInTheDocument();
    expect(screen.getByText("ui-elements")).toBeInTheDocument();
    expect(screen.getByText("transportation")).toBeInTheDocument();

    // Verify image counts are displayed
    expect(screen.getByText("5 total icons")).toBeInTheDocument();
    expect(screen.getByText("8 total icons")).toBeInTheDocument();
    expect(screen.getByText("7 total icons")).toBeInTheDocument();

    // Verify file sizes are displayed
    expect(screen.getByText("25 KB")).toBeInTheDocument();
    expect(screen.getByText("35 KB")).toBeInTheDocument();
    expect(screen.getByText("30 KB")).toBeInTheDocument();
  });

  it("filters sprite collections based on search query", () => {
    render(<SpriteCatalog {...defaultProps} searchQuery="transportation" />);

    // Should only show the transportation sprite collection
    const headers = screen.getAllByTestId("card-header");
    expect(headers.length).toBe(1);
    expect(screen.queryByText("map-icons")).not.toBeInTheDocument();
    expect(screen.queryByText("ui-elements")).not.toBeInTheDocument();
    expect(screen.getByText("transportation")).toBeInTheDocument();
  });

  it("shows no results message when search has no matches", () => {
    render(<SpriteCatalog {...defaultProps} searchQuery="nonexistent" />);
    expect(
      screen.getByText(/No sprite collections found matching "nonexistent"/i),
    ).toBeInTheDocument();

    // Should not render any cards
    const headers = screen.queryAllByTestId("card-header");
    expect(headers.length).toBe(0);
  });

  it("filters sprite collections as the user types in the search input", () => {
    const { rerender } = render(<SpriteCatalog {...defaultProps} />);
    const searchInput = screen.getByPlaceholderText("Search sprites...");

    // Simulate typing "ui" into the search box
    fireEvent.change(searchInput, { target: { value: "ui" } });

    // We rerender to simulate the parent updating searchQuery in response to user input,
    // ensuring the test reflects how the component would behave in a real app.
    expect(defaultProps.onSearchChangeAction).toHaveBeenCalledWith("ui");
    rerender(<SpriteCatalog {...defaultProps} searchQuery="ui" />);

    // Verifying only the filtered result is present ensures the UI reflects the search state,
    // not just the handler call.
    expect(screen.getByText("ui-elements")).toBeInTheDocument();
    expect(screen.queryByText("map-icons")).not.toBeInTheDocument();
    expect(screen.queryByText("transportation")).not.toBeInTheDocument();
  });

  it("calls onSearchChangeAction when search input changes", () => {
    render(<SpriteCatalog {...defaultProps} />);
    const searchInput = screen.getByPlaceholderText("Search sprites...");

    fireEvent.change(searchInput, { target: { value: "new search" } });
    expect(defaultProps.onSearchChangeAction).toHaveBeenCalledWith("new search");
  });

  it("renders download and preview buttons for each sprite collection", () => {
    render(<SpriteCatalog {...defaultProps} />);

    // We should have 3 download buttons (one for each sprite collection)
    const downloadIcons = screen.getAllByTestId("download-icon");
    expect(downloadIcons.length).toBe(3);

    // We should have 3 eye icons for preview (one for each sprite collection)
    const eyeIcons = screen.getAllByTestId("eye-icon");
    expect(eyeIcons.length).toBe(3);
  });

  it("renders preview buttons correctly", () => {
    render(<SpriteCatalog {...defaultProps} />);

    // All preview buttons should be rendered
    const previewButtons = screen.getAllByText("Preview");
    expect(previewButtons.length).toBe(3);
  });
});
