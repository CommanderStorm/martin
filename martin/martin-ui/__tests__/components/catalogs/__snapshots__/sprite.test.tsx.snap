// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SpriteCatalog Component matches snapshot for loaded state with mock data 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Sprite Catalog
    </h2>
    <div
      class="relative"
    >
      <div
        data-testid="search-icon"
      >
        Search
      </div>
      <input
        placeholder="Search sprites..."
        value=""
      />
    </div>
    <div
      class="grid"
    >
      <div
        data-testid="card-header"
      >
        <div
          data-testid="card-title"
        >
          map-icons
        </div>
        <div
          data-testid="card-description"
        >
          5 total icons
        </div>
        <div
          data-testid="card-content"
        >
          <div>
            25 KB
          </div>
          <button>
            <div
              data-testid="eye-icon"
            >
              Eye
            </div>
            Preview
          </button>
          <button>
            <div
              data-testid="download-icon"
            >
              Download
            </div>
            Download
          </button>
        </div>
      </div>
      <div
        data-testid="card-header"
      >
        <div
          data-testid="card-title"
        >
          transportation
        </div>
        <div
          data-testid="card-description"
        >
          7 total icons
        </div>
        <div
          data-testid="card-content"
        >
          <div>
            30 KB
          </div>
          <button>
            <div
              data-testid="eye-icon"
            >
              Eye
            </div>
            Preview
          </button>
          <button>
            <div
              data-testid="download-icon"
            >
              Download
            </div>
            Download
          </button>
        </div>
      </div>
      <div
        data-testid="card-header"
      >
        <div
          data-testid="card-title"
        >
          ui-elements
        </div>
        <div
          data-testid="card-description"
        >
          8 total icons
        </div>
        <div
          data-testid="card-content"
        >
          <div>
            35 KB
          </div>
          <button>
            <div
              data-testid="eye-icon"
            >
              Eye
            </div>
            Preview
          </button>
          <button>
            <div
              data-testid="download-icon"
            >
              Download
            </div>
            Download
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SpriteCatalog Component matches snapshot for loading state 1`] = `
<DocumentFragment>
  <div
    data-testid="catalog-skeleton"
  >
    <div
      data-testid="skeleton-title"
    >
      Sprite Catalog
    </div>
    <div
      data-testid="skeleton-description"
    >
      Preview all available sprite sheets and icons
    </div>
  </div>
</DocumentFragment>
`;
